File: chesslib/Position.cpp
#include "Position.h"

#include "board_utils.h"
#include "require.h"
#include "squares.h"

namespace chesslib {

Position::Position(
    square_t blackKingSquare,
    square_t whiteKingSquare,
    player_t playerToMove,
    uint32_t halfmoveClock,
    uint32_t fullmoveNumber)
    : board_(blackKingSquare, whiteKingSquare)
    , moveCounters_(halfmoveClock, fullmoveNumber)
{
    positionFlags_.setPlayerToMove(playerToMove);
}

void Position::addPiece(const PieceOnBoard& piece)
{
    board_.addPiece(piece);
}

void Position::playMove(const Move& move)
{
    const auto oldPositionFlags = positionFlags_;
    const player_t playerToMove = getPlayerToMove();
    
    const square_t originSquare = move.getOriginSquare();
    const square_t originRow = getRow(originSquare);
    const square_t destSquare = move.getDestSquare();
    piece_type_t capturedPieceType = NoPiece;

    // TODO
    if (move.isCapture())
    {
        capturedPieceType = move.getCapturedPieceType();
        if (move.isEnPassantCapture())
        {
            const auto capturedPieceSquare = encodeSquare(
                originRow, getColumn(destSquare));
            board_.erasePieceAt(capturedPieceSquare);
        }
        else {
            // TODO
            // TODO: optimization - there is no need to clear the dest square first
            board_.erasePieceAt(destSquare);
        }
        board_.updatePieceSquare(originSquare, destSquare);
    }
    else {
        board_.updatePieceSquare(originSquare, destSquare);
        if (move.isLongCastle()) {
            square_t rookOriginSquare = encodeSquare(originRow, COLUMN_A);
            square_t rookDestSquare = encodeSquare(originRow, COLUMN_D);
            board_.updatePieceSquare(rookOriginSquare, rookDestSquare);
        }
        else if (move.isShortCastle()) {
            square_t rookOriginSquare = encodeSquare(originRow, COLUMN_H);
            square_t rookDestSquare = encodeSquare(originRow, COLUMN_F);
            board_.updatePieceSquare(rookOriginSquare, rookDestSquare);
        }
    }

    if (move.isPromotion()) {
        board_.promotePawn(destSquare, move.getPromoteToPieceType());
    }

    positionFlags_.onMovePlayed();

    CastleOptions castleOptions = getCastleOptions(playerToMove);
    const piece_type_t pieceToMove = move.getPieceType();
    if (pieceToMove == King) {
        castleOptions = CastleOptions();
    }
    else if (pieceToMove == Rook) {
        if (castleOptions.isCanCastleLong() && getColumn(originSquare) == COLUMN_A) {
            castleOptions.setCanCastleLong(false);
        }
        else if (castleOptions.isCanCastleShort() && getColumn(originSquare) == COLUMN_H) {
            castleOptions.setCanCastleShort(false);
        }
    }
    positionFlags_.setCastleOptions(playerToMove, castleOptions);

    if (capturedPieceType == Rook) {
        // TODO: Optimize me
        const square_t otherBaseRow = playerToMove == Black ? 0 : MAX_ROW;
        const player_t otherPlayer = getOtherPlayer();
        CastleOptions otherCastleOptions = getCastleOptions(otherPlayer);
        if (otherCastleOptions.isCanCastleLong() && destSquare == encodeSquare(otherBaseRow, COLUMN_A)) {
            otherCastleOptions.setCanCastleLong(false);
        }
        else if (otherCastleOptions.isCanCastleShort() && destSquare == encodeSquare(otherBaseRow, COLUMN_H)) {
            otherCastleOptions.setCanCastleShort(false);
        }
        positionFlags_.setCastleOptions(otherPlayer, otherCastleOptions);
    }

    positionFlags_.setEnPassantColumn(
        move.isPawnDoubleMove() ? OptionalColumn::fromColumn(getColumn(originSquare)) : OptionalColumn());

    positionFlags_.setPlayerToMove(getOtherPlayer());

    history_.emplace_back(move, oldPositionFlags, moveCounters_);

    uint32_t newHalfmoveClock = (move.isCapture() || pieceToMove == Pawn) ? 0 : moveCounters_.getHalfmoveClock() + 1;
    uint32_t newFullmoveNumber = moveCounters_.getFullmoveNumber() + (playerToMove == Black ? 1 : 0);
    moveCounters_ = PositionMoveCounters(newHalfmoveClock, newFullmoveNumber);
}

void Position::undoMove()
{
    if (!isCanUndoMove()) {
        FAIL();
    }

    const PositionHistory& historyToUndo = history_.back();
    const player_t playerToMove = getPlayerToMove();

    const Move move = historyToUndo.getMove();
    const square_t originSquare = move.getOriginSquare();
    const square_t originRow = getRow(originSquare);
    const square_t destSquare = move.getDestSquare();

    // TODO
    if (move.isCapture())
    {
        // TODO
        board_.updatePieceSquare(destSquare, originSquare);

        const square_t capturedPieceSquare = move.isEnPassantCapture()
            ? encodeSquare(originRow, getColumn(destSquare))
            : destSquare;
        board_.addPiece({ playerToMove, move.getCapturedPieceType(), capturedPieceSquare });
    }
    else {
        board_.updatePieceSquare(destSquare, originSquare);
        if (move.isLongCastle()) {
            square_t rookOriginSquare = encodeSquare(originRow, COLUMN_A);
            square_t rookDestSquare = encodeSquare(originRow, COLUMN_D);
            board_.updatePieceSquare(rookDestSquare, rookOriginSquare);
        } else if (move.isShortCastle()) {
            square_t rookOriginSquare = encodeSquare(originRow, COLUMN_H);
            square_t rookDestSquare = encodeSquare(originRow, COLUMN_F);
            board_.updatePieceSquare(rookDestSquare, rookOriginSquare);
        }
    }

    if (move.isPromotion()) {
        board_.demoteToPawn(originSquare);
    }

    positionFlags_ = historyToUndo.getPositionFlags();
    moveCounters_ = historyToUndo.getPositionMoveCounters();

    history_.pop_back();
}

bool Position::isValid() const
{
    const player_t me = getOtherPlayer();
    const player_t attacker = getPlayerToMove();
    const square_t myKingSquare = board_.getKingSquare(me);

    if (!history_.empty()) {
        const auto parentIsInCheck = history_.back().getPositionFlags().getIsInCheck();
        if (parentIsInCheck.isEquals(false)) {
            const Move lastMove = history_.back().getMove();
            if (lastMove.getPieceType() != King) {
                const square_t originSquare = lastMove.getOriginSquare();
                const bool isMyKingBecameExposed = isSquareSlideAttackedThroughSpecificSquare(
                    originSquare, myKingSquare, board_, attacker);
                return !isMyKingBecameExposed;
            }
        }
    }

    const bool isMyKingCanBeCaptured = isSquareAttacked(myKingSquare, board_, attacker);
    return !isMyKingCanBeCaptured;
}

bool Position::isInCheck()
{
    const OptionalBoolean cachedResult = positionFlags_.getIsInCheck();
    if (cachedResult.hasValue()) {
        return cachedResult.getValue();
    }

    bool result;
    if (history_.empty()) {
        const square_t kingSquare = board_.getKingSquare(getPlayerToMove());
        result = isSquareAttacked(kingSquare, board_, getOtherPlayer());
    }
    else {
        result = isRecentMovePutsEnemyKingInCheck(history_.back().getMove());
    }

    positionFlags_.setIsInCheck(result);
    return result;
}

bool Position::isRecentMovePutsEnemyKingInCheck(Move recentMove) const
{
    const player_t otherPlayer = getOtherPlayer();
    const square_t kingSquare = board_.getKingSquare(getPlayerToMove());
    const square_t originSquare = recentMove.getOriginSquare();
    const square_t destSquare = recentMove.getDestSquare();
    if (recentMove.isCastle()) {
        const square_t rookDestColumn = (getColumn(originSquare) + getColumn(destSquare)) >> 1;
        const square_t rookDestSquare = encodeSquare(getRow(destSquare), rookDestColumn);
        return isSquareAttackedFromSpecificSquare(rookDestSquare, kingSquare, board_, otherPlayer);
    }
    else {
        return isSquareAttackedFromSpecificSquare(destSquare, kingSquare, board_, otherPlayer)
            || isSquareSlideAttackedThroughSpecificSquare(
                originSquare, kingSquare, board_, otherPlayer)
            || (recentMove.isEnPassantCapture()
                && isSquareSlideAttackedThroughSpecificSquare(
                    encodeSquare(getRow(originSquare), getColumn(destSquare)), kingSquare, board_, otherPlayer));
    }
}

namespace {

CastleOptions optimize(CastleOptions castleOptions, const Board& board, player_t player)
{
    CastleOptions result;
    const square_t row = player == Black ? MAX_ROW : 0;
    if (!board.isPlayerPiece(encodeSquare(row, COLUMN_E), player, King)) {
        return result;
    }

    result.setCanCastleLong(
        castleOptions.isCanCastleLong()
        && board.isPlayerPiece(encodeSquare(row, COLUMN_A), player, Rook));

    result.setCanCastleShort(
        castleOptions.isCanCastleShort()
        && board.isPlayerPiece(encodeSquare(row, COLUMN_H), player, Rook));

    return result;
}

}  // namespace

void Position::optimizeCastleOptions()
{
    for (player_t player = 0; player < PLAYER_COUNT; ++player) {
        setCastleOptions(player, optimize(getCastleOptions(player), board_, player));
    }
}

}

File: android/chesslibkt/src/main/java/ru/maksimbulva/chesslibkt/position/Position.kt
package ru.maksimbulva.chesslibkt.position

import ru.maksimbulva.chesslibkt.Player
import ru.maksimbulva.chesslibkt.board.*
import ru.maksimbulva.chesslibkt.getOtherPlayer
import ru.maksimbulva.chesslibkt.move.Move
import ru.maksimbulva.chesslibkt.move.MoveGenerationFilter
import ru.maksimbulva.chesslibkt.move.MoveGenerator
import ru.maksimbulva.chesslibkt.move.ScoredMove

class Position(
    blackKingSquare: Square,
    whiteKingSquare: Square,
    playerToMove: Player,
    enPassantColumn: Int?,
    castleOptions: Map<Player, CastleOptions>,
    halfmoveClock: Int,
    fullmoveNumber: Int
) {

    private var _board = Board(blackKingSquare, whiteKingSquare)
    val board: Board
        get() = _board

    private var _positionFlags = PositionFlags(0)
    val positionFlags: PositionFlags
        get() = _positionFlags

    private var _playerToMove = playerToMove
    val playerToMove: Player
        get() = _playerToMove

    private var _enPassantColumn = enPassantColumn
    val enPassantColumn: Int?
        get() = _enPassantColumn

    private var blackCastleOptions = castleOptions.getValue(Player.Black)
    private var whiteCastleOptions = castleOptions.getValue(Player.White)

    private var _halfmoveClock = halfmoveClock
    val halfmoveClock: Int
        get() = _halfmoveClock

    private var _fullmoveNumber = fullmoveNumber
    val fullmoveNumber: Int
        get() = _fullmoveNumber

    // TODO: consider another type of container if needed
    private val history = mutableListOf<PositionHistory>()

    fun clone(): Position {
        // TODO()
        return this
    }

    fun generatePseudoLegalMoves(movesFilter: MoveGenerationFilter): List<Move> {
        val movesCollection = mutableListOf<Move>()
        MoveGenerator.fillWithPseudoLegalMoves(movesCollection, movesFilter, this)
        return movesCollection
    }

    fun getCastleOptions(player: Player): CastleOptions {
        return when (player) {
            Player.Black -> blackCastleOptions
            Player.White -> whiteCastleOptions
        }
    }

    fun optimizeCastleOptions() {
//        TODO()
    }

    fun isValid(): Boolean {
//        TODO()
        return true
    }

    fun isNotValid(): Boolean = !isValid()

    fun isInCheck(): Boolean {
        TODO()
    }

    fun addPiece(piece: PieceOnBoard) {
        board.addPiece(piece)
    }

    fun playMove(move: Move) {
        val playerToMove = playerToMove
        val otherPlayer = getOtherPlayer(playerToMove)

        val originSquare = move.originSquare
        val originRow = originSquare.row
        val destSquare = move.destSquare
//        piece_type_t capturedPieceType = NoPiece;

        history.add(PositionHistory(
            move,
            _enPassantColumn,
            _positionFlags,
            _halfmoveClock,
            _fullmoveNumber
        ))

        // TODO
        if (move.isCapture) {
            TODO()
//            capturedPieceType = move.getCapturedPieceType();
            // TODO: optimization - there is no need to clear the dest square first
            // if it is not an en passant capture
            _board.erasePieceAt(move.capturedPieceSquare)
            _board.updatePieceSquare(originSquare, destSquare)
        } else {
            _board.updatePieceSquare(originSquare, destSquare)
            if (move.isCastle) {
                playCastleRookMovement(move)
            }
        }

        if (move.isPromotion) {
            _board.promotePawn(destSquare, move.promoteToPiece)
        }

        _positionFlags = _positionFlags.onMovePlayed()

//        val castleOptions = getCastleOptions(playerToMove)
//        const piece_type_t pieceToMove = move.getPieceType();
//        if (pieceToMove == King) {
//            castleOptions = CastleOptions();
//        }
//        else if (pieceToMove == Rook) {
//            if (castleOptions.isCanCastleLong() && getColumn(originSquare) == COLUMN_A) {
//                castleOptions.setCanCastleLong(false);
//            }
//            else if (castleOptions.isCanCastleShort() && getColumn(originSquare) == COLUMN_H) {
//                castleOptions.setCanCastleShort(false);
//            }
//        }
//        positionFlags_.setCastleOptions(playerToMove, castleOptions);

//        if (capturedPieceType == Rook) {
//            // TODO: Optimize me
//            const square_t otherBaseRow = playerToMove == Black ? 0 : MAX_ROW;
//            const player_t otherPlayer = getOtherPlayer();
//            CastleOptions otherCastleOptions = getCastleOptions(otherPlayer);
//            if (otherCastleOptions.isCanCastleLong() && destSquare == encodeSquare(otherBaseRow, COLUMN_A)) {
//                otherCastleOptions.setCanCastleLong(false);
//            }
//            else if (otherCastleOptions.isCanCastleShort() && destSquare == encodeSquare(otherBaseRow, COLUMN_H)) {
//                otherCastleOptions.setCanCastleShort(false);
//            }
//            positionFlags_.setCastleOptions(otherPlayer, otherCastleOptions);
//        }

        _playerToMove = getOtherPlayer(_playerToMove)

        _enPassantColumn = if (move.isPawnDoubleMove) {
            originSquare.column
        } else {
            null
        }

        _positionFlags = _positionFlags.setPlayerToMove(otherPlayer)
    }

    fun updateMoveCounters(move: Move) {
        _halfmoveClock = if (move.isCapture || move.isPawnMove) {
            0
        } else {
            _halfmoveClock + 1
        }

        if (_playerToMove == Player.White) {
            ++_fullmoveNumber
        }
    }

    fun undoMove() {
        require(history.isNotEmpty())

        val historyToUndo = history.last()
        history.removeAt(history.lastIndex)

        undoMove(historyToUndo.move)

        _playerToMove = getOtherPlayer(_playerToMove)
        _enPassantColumn = historyToUndo.enPassantColumn
        _positionFlags = historyToUndo.positionFlags
        _halfmoveClock = historyToUndo.halfmoveClock
        _fullmoveNumber = historyToUndo.fullmoveNumber
    }

    private fun undoMove(move: Move) {
        val originSquare = move.originSquare

        // TODO
        if (move.isCapture) {
            _board.updatePieceSquare(move.destSquare, originSquare)
            // TODO: optimize me, do not create temporary object
            _board.addPiece(
                PieceOnBoard(
                    playerToMove,
                    move.capturedPiece,
                    move.capturedPieceSquare
                )
            )
        } else {
            _board.updatePieceSquare(move.destSquare, originSquare)
            if (move.isCastle) {
                undoCastleRookMovement(move)
            }
        }

        if (move.isPromotion) {
            _board.demoteToPawn(originSquare)
        }
    }

    private fun playCastleRookMovement(castleMove: Move) {
        val originRow = castleMove.originSquare.row
        if (castleMove.isLongCastle) {
            _board.updatePieceSquare(
                Square(originRow, Board.COLUMN_A),
                Square(originRow, Board.COLUMN_D)
            )
        } else {
            _board.updatePieceSquare(
                Square(originRow, Board.COLUMN_H),
                Square(originRow, Board.COLUMN_F)
            )
        }
    }

    private fun undoCastleRookMovement(castleMove: Move) {
        val originRow = castleMove.originSquare.row
        if (castleMove.isLongCastle) {
            _board.updatePieceSquare(
                Square(originRow, Board.COLUMN_D),
                Square(originRow, Board.COLUMN_A)
            )
        } else {
            _board.updatePieceSquare(
                Square(originRow, Board.COLUMN_F),
                Square(originRow, Board.COLUMN_H)
            )
        }
    }
}
